src/operations/SnapshotEditor.ts

```
@@ -223,6 +223,13 @@ export class SnapshotEditor {
         if (payload && payloadName in payload) {
           warnings.push(`Encountered undefined at ${[...prefixPath, ...path].join('.')}. Treating as null`);
         }
+
+        if (this._context.getAddTypename() && payloadName === '__typename') {
+          const existingTypenameValue = deepGet(this._getNodeData(containerId), path);
```

You can skip this check - it's even more of a problem for cases where typename is missing (for some consumers of hermes, __typename is part of object identity)

---

done. For unit testing, how to test? I didn't see an existing example checking tracer warnings. I see tests for catching InvalidPayloadError. By the way, should this be an InvalidPayloadError instead of a warning? That would be more visible to consumers.

---

Yeah, failing hard here would be 👍 

---

re: unit testing the tracer, we typically fall back to testing the console output, rather than the tracer itself (since that's more stable, and these messages are likely to change). But it's also not a very proven pattern; injecting the tracer might be better.

Example test: https://github.com/convoyinc/apollo-cache-hermes/blob/master/test/unit/Cache/transactions.ts#L53-L62